test/test_util.py

```diff
@@ -44,140 +43,142 @@
 from test.lib import TestBase, with_rw_repo
 
 
-class _Member:
-    """A member of an IterableList."""
+@pytest.fixture
+def permission_error_tmpdir(tmp_path):
+    """Fixture to test permissions errors in situations where they are not overcome."""
+    td = tmp_path / "testdir"
```

Comment:
I keep choking on this as it seems more fragile than it has to be given that one could also create a tempdir with a unique name as well.

Reply:
Can you tell me more about this concern? I'll be working on `test_util.py` again soon, so I'll have a good opportunity to make further improvements. However, [`tmp_path`](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html) is already always unique: it's a `Path` object for a path like `/tmp/hw9aphg`, not `/tmp` itself.

Reply:
Ah, perfect, then there is no concern, I just wasn't aware and too overwhelmed/not very focussed at the time.

Reply:
> then there is no concern

Ah, excellent.

Some of the code this PR modifies are the tests I introduced it in #1700. Both before and after the changes here, these tests may give the mistaken impression that variables like `td` refer to objects representing paths like `/tmp/tmpdir` that are always the same, rather than paths like `/tmp/hw9aphg/tmpdir` (or, with `pytest`, a more systematic scheme but still one that makes sure to use an initially nonexistent location).

After all, if `tmp_path` really corresponds to a unique and newly created directory, why not use *it* rather than creating a `tmp_path / "testdir"` subdirectory?

I do this for several reasons:

- I think of `tmp_path` as a resource the testing framework and/or custom fixtures manage: it is created before the test and (where appropriate and feasible) removed after the test. So for testing the behavior of `util.rmtree` in deleting a directory, I make a subdirectory for that, even if I wouldn't have to.
- Currently, I have configured `pytest` not to attempt to delete the temporary directory it creates afterwards, if the test failed. This allows the directory to be examined to troubleshoot a regression that causes a failure. But this would be confusing to actually do, if there is confusion between the `pytest`-managed temporary directory that may be deleted if the test succeeds, and the directory that the code under test attempts to delete.
- In at least test, conceptually zero directories are created, and the purpose of the temporary directory is to ensure that it *doesn't* have an entry of a particular name in it, so that this nonexistent entry can itself be used, to test what happens when the code under test raises a `FileNotFoundError`.
- In some forthcoming tests I plan to propose soon (but that I've had in mind for a short while, including when working on this PR), two directories will be created. Making them subdirectories of the `tmp_path` directory will make the test code a bit simpler, as I won't have to use a more versatile but complex fixture to create multiple `pytest`-managed temporary directories. More importantly, it will allow the two directories being affected by `util.rmtree` to be distinguished from each other when examining the tests and their results, rather than having them both have meaningless names that change each time.

This may be insufficiently clear in reading the test code, and if so, I would be interested in improving the situation, perhaps by changing some variable names, adding/expanding some docstrings or comments, or both. But maybe that wouldn't help; I'm not sure.

(Regarding the broader issue of test-runner parallelism, I'll post a separate top-level comment below.)

Reply:
Thanks for the explanation! Personally I think that it's OK to leave it as is but refactor the next time one stumbles over it. Please note that I didn't read the top-level comment yet, to which I will probably reply separately.